server/src/main/java/org/opensearch/common/cache/stats/StatsHolder.java

```diff
@@ -0,0 +1,200 @@
+    private final DimensionNode statsRoot;
+
+    public StatsHolder(List<String> dimensionNames) {
+        this.dimensionNames = dimensionNames;
+        this.statsRoot = new DimensionNode(null); // The root node has no dimension value associated with it, only children
+        statsRoot.createChildrenMap();
+    }
+
+    public List<String> getDimensionNames() {
+        return dimensionNames;
+    }
+
+    // For all these increment functions, the dimensions list comes from the key, and contains all dimensions present in dimensionNames.
+    // The order has to match the order given in dimensionNames.
+    public void incrementHits(List<String> dimensionValues) {
+        internalIncrement(dimensionValues, (counter) -> counter.hits.inc(), true);
+    }
+
+    public void incrementMisses(List<String> dimensionValues) {
+        internalIncrement(dimensionValues, (counter) -> counter.misses.inc(), true);
+    }
+
+    public void incrementEvictions(List<String> dimensionValues) {
+        internalIncrement(dimensionValues, (counter) -> counter.evictions.inc(), true);
+    }
+
+    public void incrementSizeInBytes(List<String> dimensionValues, long amountBytes) {
+        internalIncrement(dimensionValues, (counter) -> counter.sizeInBytes.inc(amountBytes), true);
+    }
+
+    // For decrements, we should not create nodes if they are absent. This protects us from erroneously decrementing values for keys
+    // which have been entirely deleted, for example in an async removal listener.
+    public void decrementSizeInBytes(List<String> dimensionValues, long amountBytes) {
+        internalIncrement(dimensionValues, (counter) -> counter.sizeInBytes.dec(amountBytes), false);
+    }
+
+    public void incrementEntries(List<String> dimensionValues) {
+        internalIncrement(dimensionValues, (counter) -> counter.entries.inc(), true);
+    }
+
+    public void decrementEntries(List<String> dimensionValues) {
+        internalIncrement(dimensionValues, (counter) -> counter.entries.dec(), false);
+    }
+
+    /**
+     * Reset number of entries and memory size when all keys leave the cache, but don't reset hit/miss/eviction numbers.
+     * This is in line with the behavior of the existing API when caches are cleared.
+     */
+    public void reset() {
+        resetHelper(statsRoot);
+    }
+
+    private void resetHelper(DimensionNode current) {
+        CacheStatsCounter counter = current.getStats();
+        counter.sizeInBytes.dec(counter.getSizeInBytes());
+        counter.entries.dec(counter.getEntries());
+        if (current.hasChildren()) {
+            // not a leaf node
+            for (DimensionNode child : current.children.values()) {
+                resetHelper(child);
+            }
+        }
+    }
+
+    public long count() {
+        // Include this here so caches don't have to create an entire CacheStats object to run count().
+        return statsRoot.getStats().getEntries();
```

Comment:
The DimensionNode should expose required methods for interacting with the stats object instead of giving free access

Reply:
We could do this, but we would have to reimplement all the functions in CacheStatsCounter again in DimensionNode so StatsHolder could access them (all the getters + reset fn + new setters, for use in incrementing calls). If callers of DimensionNode already have this access to CacheStatsCounter, would it just be simpler to return the CacheStatsCounter itself?

Reply:
However we probably should implement incrementers to CacheStatsCounter, so that StatsHolder isn't manually incrementing its CounterMetrics when it increments its stats.

Reply:
We dont need to implement the getters from CacheStatsCounter, just use getters to implement the increment decrement operations we need. Essentially, StatsHolder should get everything from Dimension and should not need access or know about Counter